### ISSUE TITLE
IMapOfSets throws Stack Overflow on 'forEach' after too many 'add's

### ISSUE THREAD

**Issue by kwestlund** [marcglasberg/fast_immutable_collections#23]
[VERBOSE-2:ui_dart_state.cc(199)] Unhandled Exception: Stack Overflow
#0      FollowedByIterable.iterator (dart:_internal/iterable.dart:784:3)
#1      new FollowedByIterator (dart:_internal/iterable.dart:842:34)
#2      FollowedByIterable.iterator (dart:_internal/iterable.dart:784:35)
#3      new FollowedByIterator (dart:_internal/iterable.dart:842:34)
#4      FollowedByIterable.iterator (dart:_internal/iterable.dart:784:35)
#5      new FollowedByIterator (dart:_internal/iterable.dart:842:34)

**Comment by marcglasberg** [comment]
Would you please have some simple demonstration code I can run to see the problem?

**Comment by kwestlund** [comment]
I couldn't reproduce the above stack trace in a test snippet, but I was able to blow the stack on my machine with the following test case, and you may have to play around with the numbers to reproduce:

```
import 'package:fast_immutable_collections/fast_immutable_collections.dart';

void main() {
  const KEY_ITERS = 1000;
  const VAL_ITERS = 1000;

  try {
    var iMapOfSets = IMapOfSets.empty();
    for (int i = 0; i < KEY_ITERS; i++) {
      for (int j = 0; j < VAL_ITERS; j++) {
        iMapOfSets = iMapOfSets.add(i, j);
      }
    }
  } catch (e,s) {
    print("$e: $s");
  }
}
```

Which produces:

```
Stack Overflow: #0      MReplace.[] (package:fast_immutable_collections/src/imap/m_replace.dart:25:3)
#1      MReplace.[] (package:fast_immutable_collections/src/imap/m_replace.dart:26:55)
#2      MReplace.[] (package:fast_immutable_collections/src/imap/m_replace.dart:26:55)
#3      MReplace.[] (package:fast_immutable_collections/src/imap/m_replace.dart:26:55)

```

Thanks for the library!

**Comment by marcglasberg** [comment]
Found the problem. Instead of flushing the collections after a number of **changes**, I was flushing after a number of **reads**, which makes no sense. It's the changes which may end up creating structures so big that they may stack-overflow. After moving the flush counter to the changes, the problem is gone.

However, this is fixed only for the **sync** flush mode. The **async** flush mode only flushes after the async gap, so it can let you create structures big enough to overflow, if that's done within the current microtask. The solution to that is **tail-call-optimization**. However, Dart itself does not allow it: https://github.com/dart-lang/language/issues/1159

I have to tail-call-optimize by hand. For example, instead of this:

```
V? operator [](K key) => (key == _key) ? _value : _m[key];
```

I need to do this:
```
dynamic getVOrM(K key) => (key == _key) ? _value : _m;
dynamic containsKeyOrM(K key) => (key == _key) ? true : _m;

V? operator [](K key) {
    if ((key == _key)) {
      return _value;
    } else {
      dynamic vOrM = _m;
      while (vOrM is M) {
        vOrM = vOrM.getVOrM(key);
      }
      return vOrM as V;
    }}
```

I have actually implemented TCO for `operator []` and `containsKey()` for the `IMap`, but I'd actually would have to do that for each and every method that traverses the data tree, not only for the `IMap`, but also for the `IList` and the `ISet`. It's perhaps 2 months of work. I can't do that now.

So I have decided to remove the async flush mode, and raise a bit the `defaultFlushFactor`. The sync mode is recommended anyway, so I guess that won't make much of a difference to anyone.
